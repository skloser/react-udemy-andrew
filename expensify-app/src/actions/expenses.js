import { update } from "firebase/database";
import database, {
	ref,
	push,
	getDatabase,
	get,
	remove,
	child,
} from "../firebase/firebase";

//ADD_EXPENSE

export const addExpense = (expense) => ({
	type: "ADD_EXPENSE",
	expense,
});

export const startAddExpense = (expenseData = {}) => {
	return (dispatch) => {
		const {
			description = "",
			note = "",
			amount = 0,
			createdAt = 0,
		} = expenseData;
		const expense = { description, note, amount, createdAt };

		return push(ref(database, "expenses"), { expense }).then((expenseRef) => {
			dispatch(
				addExpense({
					id: expenseRef.key,
					...expense,
				})
			);
		});
	};
};

//REMOVE_EXPENSE

export const removeExpense = (id) => ({
	type: "REMOVE_EXPENSE",
	id,
});

export const startRemoveExpense = (id) => {
	return (dispatch) => {
		const dbRef = ref(getDatabase());
		return remove(dbRef, "expenses/" + id).then(() => {
			dispatch(removeExpense(id));
		});
	};
};

//EDIT_EXPENSE

export const editExpense = (id, updates) => ({
	type: "EDIT_EXPENSE",
	id,
	updates,
});

export const startEditExpense = (id, data) => {
	return (dispatch) => {
		const dbRef = ref(getDatabase());
		const updates = {};

		updates["/expenses/" + id] = data;

		return update(dbRef, updates).then(() => {
			dispatch(editExpense(id, data));
		});
	};
};

// SET_EXPENSES

export const setExpenses = (expenses) => ({
	type: "SET_EXPENSES",
	expenses,
});

export const startSetExpenses = () => {
	return (dispatch) => {
		const dbRef = ref(getDatabase());

		return get(child(dbRef, "expenses")).then((snapshot) => {
			const expenses = [];
			snapshot.forEach((expenseChildSnapshot) => {
				expenses.push({
					id: expenseChildSnapshot.key,
					...expenseChildSnapshot.val(),
				});
			});

			dispatch(setExpenses(expenses));
		});
	};
};
