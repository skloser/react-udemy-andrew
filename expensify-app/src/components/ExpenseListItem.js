import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
	<div>
		<p>
			<Link to={`/edit/${id}`}>Description: {description}</Link>
		</p>
		<h3>Amount: {numeral(amount / 100).format("$0,0.00")}</h3>
		<p>CreatedAt: {moment(createdAt).format("MMMM Do, YYYY")}</p>
	</div>
);

export default ExpenseListItem;
