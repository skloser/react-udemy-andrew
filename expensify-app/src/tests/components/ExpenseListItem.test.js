import React from "react";
import { shallow } from "enzyme";
import ExpenseListItem from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test("should render expense list item", () => {
	const wrapper = shallow(<ExpenseListItem expense={expenses[0]} />);
	expect(wrapper).toMatchSnapshot();
});
