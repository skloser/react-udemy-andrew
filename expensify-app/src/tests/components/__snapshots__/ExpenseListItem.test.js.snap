// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item 1`] = `
<div>
  <p>
    <Link
      replace={false}
      to="/edit/undefined"
    >
      Description: 
    </Link>
  </p>
  <h3>
    Amount: 
    $0.00
  </h3>
  <p>
    CreatedAt: 
    January 1st, 1970
  </p>
</div>
`;
